driver/makedepend.ml

```diff
@@ -96,23 +78,34 @@ let add_to_synonym_list synonyms suffix =
     Error_occurred.set ()
   end
 
+let rec is_before e e' = function
+  | [] -> raise Not_found
+  | h :: _ when h = e -> true
+  | h :: _ when h = e' -> false
+  | _ :: l -> is_before e e' l
+
 (* Find file 'name' (capitalized) in search path *)
 let find_module_in_load_path name =
-  let names = List.map (fun ext -> name ^ ext) (!mli_synonyms @ !ml_synonyms) in
-  let unames =
-    let uname = Unit_info.normalize name in
-    List.map (fun ext -> uname ^ ext) (!mli_synonyms @ !ml_synonyms)
+  let find find =
+    List.filter_map (fun ext ->
+        try Some (find (name ^ ext)) with Not_found -> None)
+      (!mli_synonyms @ !ml_synonyms)
+  in
+  let names = find Load_path.find and unames = find Load_path.find_normalized in
+  let paths = Load_path.get_paths () in
+  let rec aux current = function
+    | [] -> current
+    | f :: tl ->
+       let dir = Filename.dirname f in
+       if dir = Filename.current_dir_name then Some (dir, Filename.basename f)
+       else match current with
+            | None -> aux (Some (dir, f)) tl
+            | Some (dir', _) ->
+               if is_before dir dir' paths then aux (Some (dir, f)) tl
+               else aux current tl
```

Comment:
This complicated logic seems to be replicating what `Load_path.find` does. Can't we simply do something along the lines of
```
let names = List.map (fun ext -> name ^ ext) (!mli_synonyms @ !ml_synonyms) in
List.find_map (fun name ->
    match Load_path.find_normalized name with s -> Some s | exception Not_found -> None
  ) names 
```

Reply:
That was my first idea, but there's a problem with module shadowing. Consider the test
```
testsuite/tests/tool-ocamldep-shadowing
├── a.ml
├── a.reference
├── dir1
│   └── b.ml
└── dir2
    ├── b.mli
    └── c.mli

3 directories, 5 files
```
with the include path `-I dir1 -I dir2` and `a.ml` depending on module `B`. `Load_path` maps file names to complete file paths, keeping the entry with highest priority in the load path. It would associate `b.ml` to `dir1/b.ml` and `b.mli` to `dir2/b.mli`. Now, if we're building possible file names providing module `B`, we'll have `b.mli` first, and then `b.ml`. So, looking up for `B`, we'll find `dir2/b.mli` which is incorrect. `Load_path` doesn't account for module shadowing.

Reply:
I need a bit more time to think about this, but if the lookup semantics do not correspond to `Load_path`, why are we trying to use it here? Can we implement the necessary semantics as an alternative "lookup" operation in `Load_path` itself (that would take shadowing into account)?

Reply:
Well, the code in makedepend _almost_ a duplicate of `Load_path`, apart from the error reporting and this lookup subtleties, which is why I tried to get makedepend to use `Load_path`.
Sure, I can spend some time on this and try to get `Load_path` to understand shadowing.